### PR TITLE
General improvements to the flasher

### PR DESCRIPTION
**Pull request by HotelCalifornia**
Added a `-r` switch to the `pros flash` command so users can specify a number of times to retry failed downloads (default is to retry twice).

logic is as follows:
if an upload fails, loop until either an upload succeeds or the specified number of retries is reached.

From my testing (I was testing over VEXnet 2.0, as direct A-to-A produced no consistent problems for me), retrying just once worked the vast majority of the time. However, I did encounter an error that would occur after flashing many times (way more than in a real situation, I ~hope~ think), where `pyserial` would raise an error saying something like `The port must be configured before it can be used`. The cause of this error is a `serial.Serial()` object's `_port` property became suddenly unset. Not sure what made this happen, but the only solution I could come up with was to unplug the controller, power-cycle the cortex, and plug the controller back in (I also made the flasher say as much when it encounters this error).

Needless to say, more testing would be appreciated.

### CODE REVIEW

**Comment by HotelCalifornia** [comment]
Seems one of the issues was that telling VEXnet to go to the download channel was failing, causing the whole flash to fail.
However, when the flasher went to retry the flash, VEXnet was reporting that it was already in download mode, leading me to believe that we either 1) weren't getting a proper response from the command (i.e. the uC wasn't sending it) or 2) we weren't waiting long enough to get the response.
On a whim, I decided to change the delay between sending the download command and reading its response from 2 seconds to 3 seconds, which seems to have worked well.

**Comment by HotelCalifornia** [comment]
Didn't notice any real difference between flashing with `reset_cortex` present in `stop_user_code` and flashing without it, so I took it out.

**Comment by HotelCalifornia** [comment]
Overall, the flasher does seem more reliable than it did before, but ~some~ a lot of testing should be done to make sure.

**Comment by edjubuh** [comment]
I've attached a candidate binaries for folks to test out with. Please report back if things are working better:

~[pros_cli-2.4.3-8-g1622212-win-32bit.zip](https://github.com/purduesigbots/pros-cli/files/812485/pros_cli-2.4.3-8-g1622212-win-32bit.zip)
[pros_cli-2.4.3_8_g1622212-py3-none-any.zip](https://github.com/purduesigbots/pros-cli/files/812487/pros_cli-2.4.3_8_g1622212-py3-none-any.zip)~
Files above are stale... Use the following method:

If you have Python 3+ installed on your machine, you can use the wheel or run `pip install -e git+https://github.com/purduesigbots/pros-cli.git@flasher-tinkering#egg=pros-cli`

If you have a Windows installation, you should replace the contents of `C:\Program Files\PROS\cli` with the contents of the zip.

**Comment by HotelCalifornia** [comment]
Seems it makes a difference on Macs

File: proscli/flasher.py [hunk 1]
```diff
@@ -87,7 +90,12 @@ def flash(ctx, save_file_system, y, port, binary, no_poll):
 
     click.echo('Flashing ' + binary + ' to ' + ', '.join(port))
     for p in port:
-        prosflasher.upload.upload(p, binary, no_poll, ctx)
+        tries = 1
+        while tries <= retry:
+            code = prosflasher.upload.upload(p, y, binary, no_poll, ctx)
+            if code or code == -1000:
```
Comment (edjubuh) [inline proscli/flasher.py:98]: When does `code` return `-1000`?

File: prosflasher/upload.py [hunk 2]
```diff
@@ -140,7 +145,7 @@ def ask_sys_info(port, ctx=proscli.utils.State()):
     configure_port(port, serial.PARITY_NONE)
     debug('SYS INFO BITS: {}  PORT CFG: {}'.format(bytes_to_str(sys_info_bits), repr(port)), ctx)
     for _ in itertools.repeat(None, 10):
-        port.read(port.in_waiting)
+        port.read_all()
```
Comment (edjubuh) [inline prosflasher/upload.py]: Did you notice this helped or did it just sneak in? There are a couple other places that have `port.read(port.in_waiting)`

File: proscli/flasher.py (hunk 1 above)
Comment (HotelCalifornia) [reply proscli/flasher.py:98]: when `upload` encounters a `SerialException`

Comment (HotelCalifornia) [reply proscli/flasher.py:98]: I'm told that the `code == -1000` is redundant, but I'm thinking I'll leave it in to specify that the return value isn't only `True` or `False`

File: prosflasher/upload.py (hunk 2 above)
Comment (HotelCalifornia) [reply prosflasher/upload.py]: I looked at what `read_all()` does, and it just calls `port.read(port.in_waiting)`, so my goal in changing those calls was for consistency and cutting back on redundant code.

**Review by jwalk511 — COMMENTED** [review]
Does not work on macOS
```
My-MBP-2:plutarch codeCode$ pros flash -v -d
	DEBUG: Opening /Users/codeCode/Documents/projects/plutarch/project.pros (ProjectConfig)
Flashing /Users/codeCode/Documents/projects/plutarch/bin/output.bin to /dev/cu.usbmodem1411
Resetting cortex...	DEBUG: RESET CORTEX. PORT CFG: Serial<id=0x1037cc080, open=True>(port='/dev/cu.usbmodem1411', baudrate=115200, bytesize=8, parity='N', stopbits=1, timeout=5.0, xonxoff=False, rtscts=True, dsrdtr=False)
complete
Stopping user code... 	DEBUG: 0F 0F 21 DE 08 00 00 00 08 F1 04 

```
and hangs

**Comment by HotelCalifornia** [comment]
I can't figure out what would be causing `port.flush()` (if we go by purduesigbots/pros-atom#40) to be hanging on macOS only, nor does it seem the fix proposed in purduesigbots/pros-atom#39 has any impact on the flash, macOS or otherwise.

**Comment by HotelCalifornia** [comment]
So it turns out the flasher is and has always been broken on macOS. Neither resetting the cortex in `stop_user_code` nor closing and reopening the port seemed to work when we tested on a mac computer.

In other news, one notable issue is that VEXnet doesn't go into the download channel when we ask it to (even when we send the command 5 times, and wait 3 seconds each time for the response!!), but `ask_sys_info` reports that it is in download mode during subsequent retries, so the auto-retry feature seems to work well.

Note that changing the delay in `send_to_download_channel` from 2 seconds to 3 seconds did work for me on Windows up until two commits ago, but it's back to its old tricks.

TODO: figure out how long it really takes for VEXnet to switch to download mode, and/or add some sort of progress bar so it doesn't look like it's just hanging there.
ALSO TODO: figure out why the flasher doesn't work on macOS at all, and determine whether it's worth supporting this feature of PROS on macs in the future.

File: prosflasher/upload.py [hunk 3]
```diff
@@ -188,15 +190,15 @@ def send_to_download_channel(port, ctx=proscli.utils.State()):
         response = port.read_all()
         debug('DB CH RESPONSE: {}'.format(bytes_to_str(response)), ctx)
         response = response[-1:]
-        if response is not None and len(response) > 0 and response[0] == ACK:
+        if ask_sys_info(port, ctx, silent=True).connection_type == ConnectionType.serial_vexnet2_dl or (response is not None and len(response) > 0 and response[0] == ACK):
```
Comment (edjubuh) [inline prosflasher/upload.py]: Can you swap these around? If the response just so happens to be an ACK, then we wouldn't have to check the connection type

Comment (HotelCalifornia) [reply prosflasher/upload.py]: yeah, but if it's already in the download mode, I feel like we should finish ASAP. the function already takes 15 seconds to execute in the worst case

Comment (edjubuh) [reply prosflasher/upload.py]: I mean to do something like 
`if (response is not None and len(response) > 0 and response[0] == ACK) or ask_sys_info(port, ctx, silent=True).connection_type == ConnectionType.serial_vexnet2_dl:`

**Comment by HotelCalifornia** [comment]
Figured out that sending VEXnet into the download channel does not reliably get an ACK response, so instead we now poll the system info to see if VEXnet thinks it's in download mode, which seems to work well.

**Comment by HotelCalifornia** [comment]
After reviewing relevant documentation, it doesn't look like the `reset_cortex` function actually does anything. Will keep looking into it.

**Comment by HotelCalifornia** [comment]
[JPearman's flashing utility](https://github.com/jpearman/stm32flashCortex/blob/fc2a6291b9dec88f4e94434dda6157db0f102773/stm32.c#L315) uploads a predefined binary to perform a reset, as "the bootloader does not have a reset command"

**Comment by HotelCalifornia** [comment]
clarification: 870d3be adds the ability to flash programs when the cortex is plugged into a competition switch

**Comment by edjubuh** [comment]
Are there any remaining consistent failures that we think we can address? I know the flasher is somewhat unreliable, but if we have a definite increase in reliability, I'll call it good. @HotelCalifornia @jwalk511

**Comment by HotelCalifornia** [comment]
Well, there's the fact that `reset_cortex` doesn't really do anything, and also I don't think this flasher works on macOS.

Other than that, auto-retry + polling sysinfo to see if we're in download mode seemed to work nearly 100% of the time when I was testing (windows)-- very occasionally the port becomes unconfigured, which causes any further flashes to fail, but that is remedied either by changing USB ports and/or power-cycling the cortex/controller (I added a user-friendly message that displays in this case in ba57274).

**Comment by HotelCalifornia** [comment]
Current status:

Still trying to iron out issues on macOS. Other than that, this PR should be good to go after a bit more testing.

**Comment by HotelCalifornia** [comment]
Okay, here it is:

After d86b4d6, we've confirmed that at the very least, flashing over a direct USB A-to-A connection works from macOS, and I couldn't confirm any new problems with this change on Windows.

Looks like this should be good to merge.

**Comment by edjubuh** [comment]
We were having confirmed issues when flashing over VEXnet on Windows at least. Confirmed meaning 100% failure rate

**Comment by HotelCalifornia** [comment]
I've just run the `pros flash` command 20 times on Windows over VEXnet:

The command succeeded 19 times, and failed once (but succeeded on the retry pass).

**Comment by HotelCalifornia** [comment]
Further testing on Windows machines will be necessary

**Comment by edjubuh** [comment]
Reliability on macOS is iffy, but better if you're willing to power cycle. With programming kit: USB works most of the time, 2.0 works after a couple of attempts, 1.0 requires power cycle between successive flashes.

**Review by baylessj — APPROVED** [review]

**Pull request merged**